### PR TITLE
Refactoring: define an object for component options

### LINKED ISSUE #11053: Refactoring: define objects that contain corresponding data that are passed to the componnet
See configurator-price.component.ts component

### PR DESCRIPTION
**Pull request by Larisa-Staroverova**
Refactoring: define objects that contain corresponding data that are passed to the componnet

### CODE REVIEW

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.html [hunk 1]
```diff
@@ -2,7 +2,9 @@
   <div
     class="cx-product-card"
     [ngClass]="{
-      'cx-product-card-selected': product.selected && !singleDropdown
+      'cx-product-card-selected':
+        productCardOptions.product.selected &&
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.html]: This is probably an issue in strict mode (what if product not defined)

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.html [hunk 2]
```diff
@@ -71,13 +71,18 @@
...
-        <div class="cx-product-card-action-btn" *ngIf="!singleDropdown">
-          <ng-container *ngIf="multiSelect; else single">
+        <div
+          class="cx-product-card-action-btn"
+          *ngIf="!productCardOptions.singleDropdown"
+        >
+          <ng-container *ngIf="productCardOptions.multiSelect; else single">
             <button
-              *ngIf="product.selected; else select"
+              *ngIf="productCardOptions.product.selected; else select"
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.html]: Also issue for strict mode

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts [hunk 3]
```diff
@@ -25,11 +33,14 @@ export class ConfiguratorAttributeProductCardComponent implements OnInit {
...
   loading$ = new BehaviorSubject<boolean>(false);
 
-  @Input() preventAction = false;
-  @Input() multiSelect = false;
-  @Input() product: Configurator.Value;
-  @Input() singleDropdown = false;
-  @Input() withQuantity = true;
+  @Input()
+  productCardOptions: ConfiguratorAttributeProductCardComponentOptions = {
+    preventAction: false,
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Why do we have the default?

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts [hunk 4]
```diff
@@ -38,27 +49,33 @@ export class ConfiguratorAttributeProductCardComponent implements OnInit {
...
   ngOnInit() {
-    this.product$ = this.productService.get(this.product.productSystemId).pipe(
-      map((respProduct) => {
-        return respProduct
-          ? respProduct
-          : this.transformToProductType(this.product);
-      })
-    );
+    this.product$ = this.productService
+      .get(this.productCardOptions.product.productSystemId)
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Issue in strict mode

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts [hunk 5]
```diff
@@ -38,27 +49,33 @@ export class ConfiguratorAttributeProductCardComponent implements OnInit {
...
+            : this.transformToProductType(this.productCardOptions.product);
+        })
+      );
   }
 
   get showQuantity() {
-    return this.withQuantity && this.product.selected && this.multiSelect;
+    return (
+      this.productCardOptions.withQuantity &&
+      this.productCardOptions.product.selected &&
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Issue in strict mode

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts [hunk 6]
```diff
@@ -38,27 +49,33 @@ export class ConfiguratorAttributeProductCardComponent implements OnInit {
...
+      this.productCardOptions.multiSelect
+    );
   }
 
   onHandleSelect(): void {
     this.loading$.next(true);
-    this.handleSelect.emit(this.product.valueCode);
+    this.handleSelect.emit(this.productCardOptions.product.valueCode);
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Strict mode

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts [hunk 7]
```diff
@@ -38,27 +49,33 @@ export class ConfiguratorAttributeProductCardComponent implements OnInit {
...
   }
 
   onHandleDeselect(): void {
     this.loading$.next(true);
-    this.handleDeselect.emit(this.product.valueCode);
+    this.handleDeselect.emit(this.productCardOptions.product.valueCode);
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Strict mode

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts [hunk 8]
```diff
@@ -74,7 +91,7 @@ export class ConfiguratorAttributeProductCardComponent implements OnInit {
 
     this.handleQuantity.emit({
       quantity,
-      valueCode: this.product.valueCode,
+      valueCode: this.productCardOptions.product.valueCode,
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Strict mode

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts [hunk 9]
```diff
@@ -90,24 +107,32 @@ export class ConfiguratorAttributeProductCardComponent implements OnInit {
...
     }
     return {
-      quantity: this.product.quantity,
-      price: this.product.valuePrice,
-      priceTotal: this.product.valuePriceTotal,
-      isLightedUp: this.product.selected,
+      quantity: this.productCardOptions.product.quantity,
+      price: this.productCardOptions.product.valuePrice,
+      priceTotal: this.productCardOptions.product.valuePriceTotal,
+      isLightedUp: this.productCardOptions.product.selected,
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Issues with strict mode for the lines 124-127

**Review by ChristophHi — CHANGES_REQUESTED** [review]
I guess we need to rework that for strict mode. I did not mark all occurrences but stopped at the second file

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.html (hunk 1 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.html]: Done!

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.html (hunk 2 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.html]: Done!

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts (hunk 4 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts (hunk 5 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts (hunk 6 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts (hunk 8 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts (hunk 7 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts (hunk 9 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts (hunk 3 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Default implementation has been removed

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts [hunk 10]
```diff
@@ -25,11 +38,14 @@ export class ConfiguratorAttributeProductCardComponent implements OnInit {
...
   loading$ = new BehaviorSubject<boolean>(false);
 
-  @Input() preventAction = false;
-  @Input() multiSelect = false;
-  @Input() product: Configurator.Value;
-  @Input() singleDropdown = false;
-  @Input() withQuantity = true;
+  @Input()
+  productCardOptions: ConfiguratorAttributeProductCardComponentOptions = {
+    preventAction: false,
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Aber hier ist immer noch die default-Implementierung??

File: feature-libs/product-configurator/rulebased/components/attribute/quantity/configurator-attribute-quantity.component.ts [hunk 11]
```diff
@@ -26,24 +32,26 @@ export class ConfiguratorAttributeQuantityComponent
   quantity = new FormControl(1);
   protected sub: Subscription;
 
-  @Input() allowZero = true;
-  @Input() initialQuantity: Quantity;
-  @Input() disableQuantityActions = false;
+  @Input() quantityOptions: ConfiguratorAttributeQuantityComponentOptions = {
+    allowZero: true,
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/quantity/configurator-attribute-quantity.component.ts]: Is this default needed (as overwritten by the caller)?

File: feature-libs/product-configurator/rulebased/components/attribute/types/single-selection-bundle-dropdown/configurator-attribute-single-selection-bundle-dropdown.component.ts [hunk 12]
```diff
@@ -41,29 +47,31 @@ export class ConfiguratorAttributeSingleSelectionBundleDropdownComponent
   }
 
   ngOnInit() {
-    this.attributeDropDownForm.setValue(this.attribute.selectedSingleValue);
+    this.attributeDropDownForm.setValue(this.attribute?.selectedSingleValue);
 
-    this.selectionValue = this.attribute.values.find((value) => value.selected);
+    this.selectionValue = this.attribute?.values.find(
```
Comment (ChristophHi) [inline feature-libs/product-configurator/rulebased/components/attribute/types/single-selection-bundle-dropdown/configurator-attribute-single-selection-bundle-dropdown.component.ts]: This won't work if values is undefined

**Review by ChristophHi — CHANGES_REQUESTED** [review]

File: feature-libs/product-configurator/rulebased/components/attribute/quantity/configurator-attribute-quantity.component.ts (hunk 11 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/quantity/configurator-attribute-quantity.component.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts (hunk 10 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/product-card/configurator-attribute-product-card.component.ts]: Done!

File: feature-libs/product-configurator/rulebased/components/attribute/types/single-selection-bundle-dropdown/configurator-attribute-single-selection-bundle-dropdown.component.ts (hunk 12 above)
Comment (Larisa-Staroverova) [reply feature-libs/product-configurator/rulebased/components/attribute/types/single-selection-bundle-dropdown/configurator-attribute-single-selection-bundle-dropdown.component.ts]: Done!

**Review by ChristophHi — APPROVED** [review]

**Pull request merged**